metals/src/main/scala/scala/meta/internal/tvp/MetalsTreeViewProvider.scala

```diff
@@ -316,6 +155,232 @@ class MetalsTreeViewProvider(
     }
   }
 
+  override def onCollapseDidChange(
+      params: TreeViewNodeCollapseDidChangeParams
+  ): Unit = getTrees().foreach(_.onCollapseDidChange(params))
+
+  override def parent(
+      params: TreeViewParentParams
+  ): TreeViewParentResult =
+    params.viewId match {
+      case Project =>
+        getTrees()
+          .map(_.parent(params.nodeUri))
+          .collectFirst { case Some(value) =>
+            value
+          }
+          .getOrElse(TreeViewParentResult(null))
+      case _ => TreeViewParentResult(null)
+    }
+
+  override def onVisibilityDidChange(
+      params: TreeViewVisibilityDidChangeParams
+  ): Unit = {
+    val trees = getTrees()
+    if (params.visible) {
+      params.viewId match {
+        case TreeViewProvider.Compile =>
+          ticks(params.viewId) = sh.scheduleAtFixedRate(
+            () => {
+              val all = trees.map(_.tickBuildTreeView()).collect {
+                case Some(value) => value
+              }
+              if (all.nonEmpty) {
+                languageClient.metalsTreeViewDidChange(
+                  TreeViewDidChangeParams(all.flatten.toArray)
+                )
+              }
+            },
+            1,
+            1,
+            TimeUnit.SECONDS,
+          )
+        case TreeViewProvider.Project =>
+          val toUpdate = trees.map(_.flushPendingProjectUpdates).collect {
+            case Some(value) => value
+          }
+          if (toUpdate.nonEmpty) {
+            languageClient.metalsTreeViewDidChange(
+              TreeViewDidChangeParams(toUpdate.flatten.toArray)
+            )
+          }
+        case _ =>
+      }
+    } else {
+      ticks.remove(params.viewId).foreach(_.cancel(false))
+    }
+  }
+
+  override def onBuildTargetDidCompile(
+      id: BuildTargetIdentifier
+  ): Unit = {
+    val toUpdate = getTrees().map(_.onBuildTargetDidCompile(id)).collect {
+      case Some(value) => value
+    }
+    if (toUpdate.nonEmpty) {
+      languageClient.metalsTreeViewDidChange(
+        TreeViewDidChangeParams(toUpdate.flatten.toArray)
+      )
+    }
+  }
+}
+
+class MetalsTreeFolderViewProvider(
```

Comment:
Could you show an example of how does the tree view look like currently? With multiple folders and without?

Reply:
I think we could use a bit more granularity, but the only thing I would change is not to add the name if we only have a single one and maybe do it like `Projects (coursier)`